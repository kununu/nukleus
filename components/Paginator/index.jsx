/* eslint-disable no-plusplus */

import React, {PropTypes, Component} from 'react';

import params from 'utils/params';

import styles from './index.scss';


export default class Paginator extends Component {
  static propTypes = {
    baseLink: PropTypes.element.isRequired,
    pathname: PropTypes.string.isRequired,
    query: PropTypes.object,
    queryKey: PropTypes.string.isRequired,
    totalPages: PropTypes.number.isRequired
  };

  static defaultProps = {
    pathname: '/',
    query: {},
    queryKey: 'page'
  };

  getPageRange (page) {
    let counter = 0;
    let currentPage = page - 3;
    const {totalPages} = this.props;
    const pageRange = [];

    for (let i = currentPage; i <= totalPages; i++) {
      if (currentPage <= 0) {
        currentPage++;
      } else if (counter < 7) {
        pageRange.push(currentPage);
        currentPage++;
        counter++;
      }
    }

    return pageRange;
  }

  getNewProps (queryValue) {
    const {
      baseLink,
      pathname,
      query,
      queryKey
    } = this.props;

    if (baseLink.props.to) {
      return {
        to: {
          pathname,
          query: {...query, [queryKey]: queryValue}
        }
      };
    }

    const location = baseLink.props.href ? 'href' : 'path';

    return {
      [location]: `${pathname}?${params({...query, [queryKey]: queryValue})}`
    };
  }

  render () {
    const {
      baseLink,
      totalPages,
      queryKey,
      query
    } = this.props;

    const currentPage = Number(query[queryKey]) || 1;
    const totalPagesArray = this.getPageRange(currentPage, totalPages);
    const previousPage = currentPage !== 1 ? currentPage - 1 : currentPage;
    const nextPage = currentPage !== totalPages ? currentPage + 1 : currentPage;

    return (
      <div className={styles.paginator}>
        <ul className={styles.paginatorList} role="navigation" aria-label="Pagination">
          <li className={(totalPages === 1 || !currentPage || currentPage === 1) && styles.disabled}>
            {React.cloneElement(baseLink, this.getNewProps(previousPage), '<')}
          </li>

          {totalPagesArray.map(item =>
            <li key={item} className={currentPage === item && styles.active}>
              {React.cloneElement(baseLink, this.getNewProps(item), item)}
            </li>
          )}

          <li className={currentPage === totalPages && styles.disabled}>
            {React.cloneElement(baseLink, this.getNewProps(nextPage), '>')}
          </li>
        </ul>
      </div>
    );
  }
}
