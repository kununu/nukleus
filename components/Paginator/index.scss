@import '~kununu-theme/scss/base/variables';

.paginatorList {
  display: inline-block;
  padding-left: 0;
  border-radius: $border-radius-base;

  * {
    box-sizing: border-box;
  }

  > li {
    display: inline; // Remove list-style and block-level defaults

    > a {
      position: relative;
      float: left;
      padding: $padding-base-vertical $padding-base-horizontal;
      line-height: $line-height-base;
      text-decoration: none;
      color: $pagination-color;
      background-color: $pagination-bg;
      border: 1px solid $pagination-border;
      margin-left: -1px;
      transition: color .25s;

      &:hover,
      &:focus {
        z-index: 2;
        color: $pagination-hover-color;
        background-color: $pagination-hover-bg;
        border-color: $pagination-hover-border;
      }
    }

    &:first-child {
      > a {
        margin-left: 0;
        border-bottom-left-radius: $border-radius-base;
        border-top-left-radius: $border-radius-base;
      }
    }

    &:last-child {
      > a {
        border-bottom-right-radius: $border-radius-base;
        border-top-right-radius: $border-radius-base;
      }
    }

    &.active {
      > a {
        &,
        &:hover,
        &:focus {
          z-index: 3;
          color: $pagination-active-color;
          background-color: $pagination-active-bg;
          border-color: $pagination-active-border;
          cursor: default;
          pointer: disabled;
          pointer-events: none;
        }
      }
    }

    &.disabled {
      > a {
        &,
        &:hover,
        &:focus {
          pointer: disabled;
          pointer-events: none;
          color: $pagination-disabled-color;
          background-color: $pagination-disabled-bg;
          border-color: $pagination-disabled-border;
          cursor: $cursor-disabled;
        }
      }
    }
  }
}

@media (max-width: $screen-xs-min) {
  .paginatorList > li > a {
    padding: 11px;
  }
}
