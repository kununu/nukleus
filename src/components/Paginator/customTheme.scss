@import '~@kununu/kununu-theme-v2/scss/base/variables';

.paginatorList {
  display: inline-block;
  padding-left: 0;

  * {
    box-sizing: border-box;
  }

  > li {
    display: inline-block;
    vertical-align: middle;

    > a {
      position: relative;
      float: left;
      padding: $padding-base-vertical $padding-base-horizontal;
      line-height: $line-height-base;
      text-decoration: none;
      color: $blue;
      background-color: $white;
      margin-left: -1px;
      transition: color .25s;

      &:hover,
      &:focus {
        z-index: 2;
        font-weight: bold;
      }
    }

    &:first-child {
      > a {
        margin-left: 0;
        border-bottom-left-radius: $border-radius-base;
        border-top-left-radius: $border-radius-base;
      }
    }

    &:last-child {
      > a {
        border-bottom-right-radius: $border-radius-base;
        border-top-right-radius: $border-radius-base;
      }
    }

    &.pageActive {
      > a {
        &,
        &:hover,
        &:focus {
          z-index: 3;
          cursor: default;
          pointer: disabled;
          pointer-events: none;
          color: $blue;
          border: 1px solid $green;
          border-radius: 100px;
          height: 30px;
          width: 30px;
          text-align: center;
          padding: 3px 0 0;
        }
      }
    }

    &.pageDisabled {
      > a {
        &,
        &:hover,
        &:focus {
          pointer: disabled;
          pointer-events: none;
          color: $gray-light;
          background-color: $white;
          border-color: #ddd;
          cursor: not-allowed;
        }
      }
    }
  }
}