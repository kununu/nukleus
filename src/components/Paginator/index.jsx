/* eslint-disable no-plusplus */

import React from 'react';
import PropTypes from 'prop-types';
import queryString from 'query-string';

import styles from './index.scss';

export default class Paginator extends React.Component {
  static propTypes = {
    baseLink: PropTypes.element.isRequired,
    pathname: PropTypes.string,
    query: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.object,
    ]),
    queryKey: PropTypes.string,
    totalPages: PropTypes.number.isRequired,
  };

  static defaultProps = {
    pathname: '/',
    query: {},
    queryKey: 'page',
  };

  getPageRange (page) {
    let counter = 0;
    let currentPage = page - 3;
    const {totalPages} = this.props;
    const pageRange = [];

    for (let i = currentPage; i <= totalPages; i++) {
      if (currentPage <= 0) {
        currentPage++;
      } else if (counter < 7) {
        pageRange.push(currentPage);
        currentPage++;
        counter++;
      }
    }

    return pageRange;
  }

  getNewProps (queryValue) {
    const {
      baseLink,
      pathname,
      query,
      queryKey,
    } = this.props;

    const queryObject = this.parseQueryObject(query);
    const queryParams = queryString.stringify({...queryObject, [queryKey]: queryValue});

    if (baseLink.props.to) {
      return {
        to: {
          pathname,
          search: queryParams,
        },
      };
    }

    const location = baseLink.props.href ? 'href' : 'path';

    return {[location]: `${pathname}?${queryParams}`};
  }

  parseQueryObject = query => (typeof query === 'object') ? query : queryString.parse(query);

  render () {
    const {
      baseLink,
      query,
      queryKey,
      totalPages,
    } = this.props;

    const queryObject = this.parseQueryObject(query);

    const currentPage = Number(queryObject[queryKey]) || 1;
    const totalPagesArray = this.getPageRange(currentPage);
    const previousPage = currentPage !== 1 ? currentPage - 1 : currentPage;
    const nextPage = currentPage !== totalPages ? currentPage + 1 : currentPage;

    return (
      <div className={styles.paginator}>
        <ul
          className={styles.paginatorList}
          role="navigation"
          aria-label="Pagination"
        >
          <li className={(totalPages === 1 || !currentPage || currentPage === 1) ? styles.disabled : undefined}>
            {React.cloneElement(baseLink, this.getNewProps(previousPage), '<')}
          </li>

          {totalPagesArray.map(item => (
            <li
              key={item}
              className={currentPage === item ? styles.active : undefined}
            >
              {React.cloneElement(baseLink, this.getNewProps(item), item)}
            </li>
          ))}

          <li className={currentPage === totalPages ? styles.disabled : undefined}>
            {React.cloneElement(baseLink, this.getNewProps(nextPage), '>')}
          </li>
        </ul>
      </div>
    );
  }
}
