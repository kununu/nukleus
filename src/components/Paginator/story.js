import React from 'react';
import {storiesOf} from '@storybook/react';
import {Link} from 'react-router-dom';
import {defaultOptions} from 'defaultOptions';
import {withOptions} from '@storybook/addon-options';
// import {action} from '@storybook/addon-actions';
// import {withNotes} from '@storybook/addon-notes';
// import {withInfo} from '@storybook/addon-info';
// import {withKnobs, text, boolean} from '@storybook/addon-knobs/react';
// import backgrounds from '@storybook/addon-backgrounds';

import Paginator from './index';

storiesOf('Paginator', module)
  .addDecorator(withOptions({...defaultOptions}))
  .add('with some emoji', () => (
    <Paginator
      totalPages={10}
      pathname="/"
      query={{}}
      baseLink={<a href="/">hi</a>} />
  ));
