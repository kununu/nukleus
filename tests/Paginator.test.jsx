import React from 'react';
import {Link} from 'react-router';
import renderer from 'react-test-renderer';
import Paginator from 'Paginator'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies

test('Renders Paginator with react-router links without crashing', () => {
  const component = renderer.create(
    <Paginator
      totalPages={10}
      pathname="test"
      queryKey="pagination"
      baseLink={<Link to={{pathname: '/test/'}}>1</Link>} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders Paginator with react-router links without crashing', () => {
  const component = renderer.create(
    <Paginator
      totalPages={10}
      query={{test: 'react-router'}}
      pathname="test"
      queryKey="pagination"
      baseLink={<a href="/test/">1</a>} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
