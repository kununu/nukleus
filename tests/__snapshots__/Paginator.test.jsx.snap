exports[`test Renders Paginator with react-router links without crashing 1`] = `
<div
  className="paginator">
  <ul
    aria-label="Pagination"
    className="pagination"
    role="navigation">
    <li
      className="disabled pointer-disabled">
      <a
        onClick={[Function]}
        style={Object {}}>
        <
      </a>
    </li>
    <li
      className="active pointer-disabled">
      <a
        onClick={[Function]}
        style={Object {}}>
        1
      </a>
    </li>
    <li
      className="inActive">
      <a
        onClick={[Function]}
        style={Object {}}>
        2
      </a>
    </li>
    <li
      className="inActive">
      <a
        onClick={[Function]}
        style={Object {}}>
        3
      </a>
    </li>
    <li
      className="inActive">
      <a
        onClick={[Function]}
        style={Object {}}>
        4
      </a>
    </li>
    <li
      className="inActive">
      <a
        onClick={[Function]}
        style={Object {}}>
        5
      </a>
    </li>
    <li
      className="inActive">
      <a
        onClick={[Function]}
        style={Object {}}>
        6
      </a>
    </li>
    <li
      className="inActive">
      <a
        onClick={[Function]}
        style={Object {}}>
        7
      </a>
    </li>
    <li
      className={false}>
      <a
        onClick={[Function]}
        style={Object {}}>
        >
      </a>
    </li>
  </ul>
</div>
`;

exports[`test Renders Paginator with react-router links without crashing 2`] = `
<div
  className="paginator">
  <ul
    aria-label="Pagination"
    className="pagination"
    role="navigation">
    <li
      className="disabled pointer-disabled">
      <a
        href="test?test=react-router&pagination=1">
        <
      </a>
    </li>
    <li
      className="active pointer-disabled">
      <a
        href="test?test=react-router&pagination=1">
        1
      </a>
    </li>
    <li
      className="inActive">
      <a
        href="test?test=react-router&pagination=2">
        2
      </a>
    </li>
    <li
      className="inActive">
      <a
        href="test?test=react-router&pagination=3">
        3
      </a>
    </li>
    <li
      className="inActive">
      <a
        href="test?test=react-router&pagination=4">
        4
      </a>
    </li>
    <li
      className="inActive">
      <a
        href="test?test=react-router&pagination=5">
        5
      </a>
    </li>
    <li
      className="inActive">
      <a
        href="test?test=react-router&pagination=6">
        6
      </a>
    </li>
    <li
      className="inActive">
      <a
        href="test?test=react-router&pagination=7">
        7
      </a>
    </li>
    <li
      className={false}>
      <a
        href="test?test=react-router&pagination=2">
        >
      </a>
    </li>
  </ul>
</div>
`;
